### PR TITLE
Added list all items in a shopcart endpoint

### PR DESCRIPTION
**Pull request by RogerHuangPKX**
**Made the following changes:**

- update_shopcart_item
-  list_shopcart_items

### CODE REVIEW

**Comment by manisha-goyal** [comment]
Please add test cases for all the new routes added.

File: service/routes.py [hunk 1]
```diff
@@ -157,7 +173,40 @@ def update_shopcarts(shopcart_id):
...
+@app.route("/shopcarts/<int:shopcart_id>/items/<int:item_id>", methods=["PUT"])
+def update_shopcart_item(shopcart_id, item_id):
+    """
+    Update a Shopcart Item
+
+    This endpoint will update an Shopcart Item based the body that is posted
+    """
+    app.logger.info("Request to update shopcart item with id: %s", item_id)
+    check_content_type("application/json")
+    import time
```
Comment (manisha-goyal) [inline service/routes.py:185]: Not necessary to track time taken on the request

File: service/routes.py [hunk 2]
```diff
@@ -142,7 +142,23 @@ def update_shopcarts(shopcart_id):
...
+    Retrieve all items in a Shopcart
+    """
+    import time
+    start_time=time.time()
+    app.logger.info("Request to list items in a shopcart with id [%s]", shopcart_id)
+
+    # Attempt to find the Shopcart and abort if not found
+    shopcart = Shopcart.find(shopcart_id)
+    if not shopcart:
+        abort(status.HTTP_404_NOT_FOUND, f"Shopcart with id '{shopcart_id}' was not found.")
```
Comment (manisha-goyal) [inline service/routes.py:157]: Use error(status_code, reason) helper method instead of calling abort to enable logging.

File: service/routes.py [hunk 3]
```diff
@@ -157,7 +173,40 @@ def update_shopcarts(shopcart_id):
...
+    This endpoint will update an Shopcart Item based the body that is posted
+    """
+    app.logger.info("Request to update shopcart item with id: %s", item_id)
+    check_content_type("application/json")
+    import time
+    start_time=time.time()
+    # See if the shopcart exists and abort if it doesn't
+    shopcart = Shopcart.find(shopcart_id)
+    if not shopcart:
+        abort(
```
Comment (manisha-goyal) [inline service/routes.py:190]: Use error(status_code, reason) helper method instead of calling abort to enable logging.

File: service/routes.py [hunk 4]
```diff
@@ -157,7 +173,40 @@ def update_shopcarts(shopcart_id):
...
+            status.HTTP_404_NOT_FOUND,
+            f"Shopcart with id '{shopcart_id}' was not found.",
+        )
+
+    # See if the item exists and abort if it doesn't
+    item = ShopcartItem.find(item_id)
+    if not item:
+        abort(
```
Comment (manisha-goyal) [inline service/routes.py:198]: Use error(status_code, reason) helper method instead of calling abort to enable logging.

**Comment by JiaxiXie** [comment]
Maybe we can modify the pull request?

**Comment by manisha-goyal** [comment]
> Maybe we can modify the pull request?

Yes I agree. @RogerHuangPKX can you please split this pull request and create a separate one for each feature.

File: tests/test_routes.py [hunk 5]
```diff
@@ -161,6 +161,18 @@ def test_update_shopcart__with_invalid_id(self):
     #####################################################################
 
     # Todo: Add shopcart item test cases here
+def test_list_all_items(self):
```
Comment (manisha-goyal) [inline tests/test_routes.py:164]: Please fix alignment. Make sure the tests are passing.

**Review by manisha-goyal — CHANGES_REQUESTED** [review]

File: service/routes.py (hunk 2 above)
Comment (manisha-goyal) [inline service/routes.py:157]: Please use the error helper method for logging instead of directly aborting.

**Review by manisha-goyal — CHANGES_REQUESTED** [review]

**Comment by RogerHuangPKX** [comment]
Please review the newest code with successful testing.

File: tests/test_routes.py [hunk 6]
```diff
@@ -160,7 +162,23 @@ def test_update_shopcart__with_invalid_id(self):
...
+        test_shopcart.create()
+        items = ShopcartItemFactory.create_batch(3, shopcart_id=test_shopcart.id)
+        for item in items:
+            item.create()
+            test_shopcart.items.append(item)
+        test_shopcart.update()
+        # Ensure the shopcart has items
+        self.assertEqual(len(test_shopcart.items), 3)
+        # List all items
+        response = self.client.get(BASE_URL / {test_shopcart.id} / items)
```
Comment (manisha-goyal) [inline tests/test_routes.py:178]: >       response = self.client.get(BASE_URL / {test_shopcart.id} / items)
E       TypeError: unsupported operand type(s) for /: 'str' and 'set'

tests/test_routes.py:178: TypeError


Please fix this error

**Review by manisha-goyal — CHANGES_REQUESTED** [review]

File: tests/test_routes.py (hunk 6 above)
Comment (manisha-goyal) [reply tests/test_routes.py:178]: Correct code:

response = self.client.get(f"{BASE_URL}/{test_shopcart.id}/items")

Comment (RogerHuangPKX) [reply tests/test_routes.py:178]: fixed

Comment (manisha-goyal) [reply tests/test_routes.py:178]: The code is still the same, did you forgot to push your change?

**Review by manisha-goyal — CHANGES_REQUESTED** [review]

**Review by manisha-goyal — APPROVED** [review]

**Pull request merged**